README.md
# NAME

Datahub::Factory - A CLI tool that transforms and transports data from a data source to a data sink.

# STATUS

[![Build Status](https://travis-ci.org/thedatahub/Datahub-Factory.svg?branch=master)](https://travis-ci.org/thedatahub/Datahub-Factory)
[![CPANTS kwalitee](https://cpants.cpanauthors.org/dist/Datahub-Factory.png)](https://cpants.cpanauthors.org/dist/Datahub-Factory)

# SYNOPSIS

    # From the command line
    $ dhconveyor <command> OPTIONS

    # Transport data via a pipeline configuration file, without output
    $ dhconveyor transport -p pipeline.ini

    # Pretty output
    $ dhconveyor transport -p pipeline.ini -v

    # Show logging output
    # Log levels: 1 - 3
    $ dhconveyor transport -p pipeline.ini -L 3

    # Only process the first 5 records
    $ dhconveyor transport -p pipeline.ini -n 5

    # Breaking up the pipeline configuration file in separate files
    $ dhconveyor transport -g general.ini -i importer.ini -f fixer.ini -e exporter.ini

    # Pushing a JSON file to a search index (Solr)
    $ dhconveyor index -p solr.ini

    # Pretty output
    $ dhconveyor index -p solr.ini -v

    # Show logging output
    # Log levels: 1 - 3
    $ dhconveyor index -p solr.ini -L 3

# DESCRIPTION

This package implements a command line ETL (Extract, Transform, Load) toolkit written as a wrapper around [Catmandu](https://metacpan.org/pod/Catmandu) and its ecosystem of Perl modules.

Features:

- Configuration files as ETL pipelines. The [Catmandu](https://metacpan.org/pod/Catmandu) command takes input via CLI options and arguments that are defined in its modules. Depending on the invocation, commands can end up with a long list of parameters containing potentially sensitive information. Sequestering that information in separate files allows users to approach pipelines as a configuration management concern.
- Conditional transforming. Pipelines can define multiple Catmandu Fixes. A check on a context dependent value (i.e. a repository field) allows the toolkit to dynamically apply the correct fix at runtime.
- Loose coupling with the Catmandu ecosystem. Wrapping Catmandu modules brings dependency inversion. This makes it easier to swap out Catmandu modules for something else without touching the infrastructure configuration.
- Robust processing with an increased fault-tolerance. Invalid records or input will simply be logged, rather then halting the entire process.
- Extensibility. Leveraging a modular approach, this toolkit can be expanded by custom modules for specific use cases.

[Datahub::Factory](https://metacpan.org/pod/Datahub::Factory) fetches data from several sources as specified by the Importer settings, executes a [Catmandu::Fix](https://metacpan.org/pod/Catmandu::Fix) and sends it to a data sink, set via an Exporter. Several importer and exporter modules are provided out of the box, but developers can extend the functionality with their own modules.

[Datahub::Factory](https://metacpan.org/pod/Datahub::Factory) supports [Log4perl](https://metacpan.org/pod/Log4perl).

# USE

## Command line options

All commands share the following switches:

- --log\_level --L \[int\]

    Set the log\_level. Takes a numeric parameter. Supported levels are: 1 (WARN), 
    2 (INFO), 3 (DEBUG). WARN (1) is the default.

- --log\_output

    Selects an output for the log messages. By default, it will send them to STDERR 
    (pass STDERR as parameter), but STDOUT (STDOUT) and a log file.

- --verbose -v

    Set verbosity. Invoking the command with the --verbose, -v flag will render 
    verbose output to the terminal.

- --number -n \[int\]

    Set number of records to process. Invoking the transport command with the --number, -n flag will process the first \[int\] records instead of all records available at the data source.

## Available Commands

### help COMMAND

Documentation about command line options.

### [transport OPTIONS](https://metacpan.org/pod/Datahub::Factory::Command::transport)

Fetch data from a local or remote source, convert the data to a target format and structure and export the data to a local or remote data sink.

### [index OPTIONS](https://metacpan.org/pod/Datahub::Factory::Command::index)

Fetch data from a local source, and push it to an enterprise search engine in bulk. Currently only supports Apache Solr ([https://lucene.apache.org/solr/](https://lucene.apache.org/solr/))

# CONFIGURATION

Pipelines are defined in configuration files which are formatted according to the INI structure as expected by the [Config::Simple](https://metacpan.org/pod/Config::Simple) library. Any pipeline consists of 4 parts: a General block, an Importer block, a Fixer block and an Exporter block.

Examples can be found in [https://github.com/thedatahub/Datahub-Factory-Pipelines](https://github.com/thedatahub/Datahub-Factory-Pipelines).

A simple example that pushes OAI data to a YAML output on STDOUT:

    [General]
    id_path = administrativeMetadata.recordWrap.recordID.0._

    [Importer]
    plugin = OAI

    [plugin_importer_OAI]
    endpoint =  https://datahub.vlaamsekunstcollectie.be/oai
    handler = +Catmandu::Importer::OAI::Parser::lido
    metadata_prefix = oai_lido

    [Fixer]
    plugin = Fix

    [plugin_fixer_Fix]
    file_name = '/home/foobar/datahub.fix'

    [Exporter]
    plugin = YAML

    [Exporter_YAML]

Note: The datahub.fix file is required, but can be left empty.

An example defining multiple fix transforms based on a context dependent value:

    [General]
    id_path = 'administrativeMetadata.recordWrap.recordID.0._'

    [Importer]
    plugin = OAI

    [plugin_importer_OAI]
    # endpoint = 'http://collections.britishart.yale.edu/oaicatmuseum/OAIHandler'
    endpoint = https://datahub.vlaamsekunstcollectie.be/oai
    handler = +Catmandu::Importer::OAI::Parser::lido
    metadata_prefix = oai_lido

    [Fixer]
    plugin = Fix

    [plugin_fixer_Fix]
    condition_path = '_metadata.administrativeMetadata.0.recordWrap.recordSource.0.legalBodyName.0.appellationValue.0._'
    fixers = MSK, GRO

    [plugin_fixer_GRO]
    condition = 'Musea Brugge - Groeningemuseum'
    file_name = '/Users/foobar/groeninge.fix'
    
    [plugin_fixer_MSK]
    condition = 'Museum voor Schone Kunsten Gent'
    file_name = '/Users/foobar/msk.fix'

    [Exporter]
    plugin = YAML

    [plugin_exporter_YAML]

Note: condition\_path contains the Fix path to the node that contains the context-dependent value. The condtion parameter in each fixer contains the value against which the conditional check is performed.

# API

Datahub::Factory leverages a plugin-based architecture. This makes extending the toolkit with new functionality fairly trivial.

New commands can be added by creating a new, separate Perl module that contains a \`command\_name.pm\` file in the \`lib/Datahub/Factory/Command\` path. 
Datahub::Factory uses the [Datahub::Factory::Command](https://metacpan.org/pod/Datahub::Factory::Command) namespace and leverages [App::Cmd](https://metacpan.org/pod/App::Cmd) internally.

New [Datahub::Factory::Importer](https://metacpan.org/pod/Datahub::Factory::Importer), [Datahub::Factory::Exporter](https://metacpan.org/pod/Datahub::Factory::Exporter), [Datahub::Factory::Fixer](https://metacpan.org/pod/Datahub::Factory::Fixer), [Datahub::Factory::Indexer](https://metacpan.org/pod/Datahub::Factory::Indexer) plugins can be added in the same way.

# AUTHORS

- Matthias Vandermaesen `matthias.vandermaesen@vlaamsekunstcollectie.be`
- Pieter De Praetere `pieter@packed.be`

# COPYRIGHT AND LICENSE

This software is copyright (c) 2016, 2019 by PACKED, vzw, Vlaamse Kunstcollectie, vzw.

This is free software; you can redistribute it and/or modify it under the terms of the GNU General Public License, Version 3, June 2007.


Changes
Revision history for Datahub-Factory

{{$NEXT}}

1.77  2019-07-19 15:53:48 CEST

    - Fixed: Add missing LWP::Protocol::https.

1.76  2019-07-18 12:27:09 CEST

    - Removed: stale TMS importer module.
    - Fixed: missing --number documentation in README.
    - Fixed: Add missing Module::Build to make Travis happy.
    - Fixed: Added missing prereq as suggested by CPANTS.

1.75  2019-07-15 17:09:34 CEST

    - Fixed: Error output for Solr indexer without Perl errors.
    - Added: missing Catmandu::DBI dependency
    - Fixed: all errors now prefixed with item_id for improved logging. 

1.74  2018-09-03 14:11:03 CEST

    - Added: Datahub::Factory::FixFileNotFound error.
    - Added: Datahub::Factory::InvalidCondition error.
    - Fixed: fixes #62 fixer Fix throws an error file_name is undefined
    - Added: fixes #61 break after finite number of records in transport.
    - Fixed: issue #60 concatenation error when a HTTP error is thrown.
    - Fixed: issue #59 Error when adding only one fixer when Fix plugin is used.
    - Added: support display parameter for Catmandu::CA.

1.73  2017-11-30 12:03:28 CET

    - Changed: added a General block to the configuration.
    - Changed: added splitted configuration files.
    - Fixed: indexer throws an ugly perl error for Catmandu::HTTPError.
    - Changed: generate README.md from separate POD documentation.
    - Proposed fix for missing prereqs as reported by CPANTS.
    - Fixed #51: Non-existing pipeline.ini should return proper error message. 
    - Added: documentation for the Datahub::Factory::Flash command.
    - Added: documentation for the Index command.

1.72  2017-10-06 15:19:32 CEST

    - Changed: removed conf folder.
    - Changed: removed documentation folder.
    - Added: .perltidyrc and .tidyallrc
    - Changed: flexible configuration based on a local datahubfactory.yml file.
    - Changed: fixes #48 Log4perl throws forgot to call init() error.
    - Changed: processing the output of the Solr indexer. Added error handling.
    - Added: dhconveyor index command with a basic Solr indexer.
    - Changed: Pipeline configuration is now polymorphic to cater different pipeline formats.
    - Fixed: missing Log::Any::Adapter::Log4perl dependency.

1.71  2017-09-06 10:54:08 CEST

    - Fixed: Licensing via GPv3 on milla release.
    - Fixed: missing Moose::Role.

1.70  2017-09-05 14:03:04 CEST

    - Changed: improved milla / dzil release flow.
    - Added: missing dependencies for testing in cpanfile.
    - Added: first iteration for .travis.yml file.
    - Changed: Clean up use imports for all modules.
    - Changed: Moved dhconveyor executable from script/ to bin/ directory for Module::Build.
    - Changed: Use Module::Build instead of Module::Build::Tiny.
    - Changed: Removed non-used dependencies which should be resolved automatically.
    - Changed: Upgraded Catmandu 1.0602 to 1.0603.

1.6.1  2017-09-04 09:11:34 CEST

    - Fixed: faulty references to deprecated Datahub::Factory::PipelineConfig.

1.6  2017-08-29 17:00:28 CEST

    - Changed: condition property renamed to explicit condition_path.
    - Changed: id_path configuration now done in importer instead of fixer.
    - Changed: All validation of the pipeline consolidated in Pipeline module.
    - Changed: refactored the Condition module.
    - Changed: refactored the Pipeline module.
    - Changed: removed deprecated legacy CLI options from the Pipeline

1.5  2017-08-27 14:38:46 CEST

    - Updated: Catmandu dependency from 1.0304 to 1.0602.
    - fixed: uninitialized error on item_id.
    - fixed: missing add method in stock datahub module.
    - added: verbose output instead of relying on stock log output.
    - changed: refactored the transport command.
    - removed: deprecated cli options and validation.

1.4  2017-07-08 13:40:43 CEST
    - added: JSON importer per issue ##43
    - changed: moved fetching id from item before item is fixed.
    - added: JSON exporter.
    - Fixed #33: Datahub::Factory::Fixer::Condition breaks with too many file opens
    - changed: display counter number in debug information
    - changed: confrom to new Catmandu::Store::Datahub API
    - added: missing dependency Catmandu::Importer::XML
    - fixed: string check condition

1.3  2017-05-12 12:45:35 UTC
	- Improved error handling: try/catch and Catmandu::Error->throw

1.2  2017-05-11 11:48:02 UTC
        - Updated documentation
        - Refactor of Exporter.pm

1.1  2017-05-11 09:27:31 UTC
        - Refactor of conditional Fix modules
        - Updated documentation

1.0  2017-05-09 13:50:42 UTC
        - Updated version number

0.11  2017-05-09 13:42:32 UTC
	- Updated documentation

0.10  2017-04-06 07:12:16 UTC
        - Added AdlibOAI

0.9  2017-03-27 11:15:53 UTC
        - Added a configurable logger
        - Updated error handling
        - Updated documentation

0.8  2017-03-22 14:54:08 UTC
	- Refactoring to remove unused scripts and branches
	- Refactoring to loosely couple importer/fixer/exporter
	- Updated documentation

0.7  2017-03-20 12:54:37 UTC
        - Moved all institution-specific importers (MSK, KMSKA, VKC) to separate modules

0.6  2017-03-20 11:03:57 UTC
	- Removed unused Switch import

0.5  2017-02-22 11:09:26 UTC
        - Added Importer::Datahub
        - Added merge command

0.03  2017-02-17 09:09:06 UTC
	- Configuratble id_path for transport.pm
        - Added Importer::OAI

0.02  2017-02-15 10:13:05 UTC
        - Refactor to use scripts/dhconveyor
        - Refactor to use App::Cmd
        - Updated logging
        - Refactor to use configuration file
        - Refactor to use pipeline configuration file

0.01  2017-01-03 10:42:51 UTC
	- Bugfix
        - Initial release


Build.PL

# This file was automatically generated by Dist::Zilla::Plugin::ModuleBuild v6.010.
use strict;
use warnings;

use Module::Build 0.28;


my %module_build_args = (
  "build_requires" => {
    "Module::Build" => "0.28"
  },
  "configure_requires" => {
    "Module::Build" => "0.28"
  },
  "dist_abstract" => "A conveyor belt which transports data from a data source to a data sink.",
  "dist_author" => [
    "Matthias Vandermaesen <matthias.vandermaesen\@vlaamsekunstcollectie.be>"
  ],
  "dist_name" => "Datahub-Factory",
  "dist_version" => "1.76",
  "license" => "gpl",
  "module_name" => "Datahub::Factory",
  "recursive_test_files" => 1,
  "requires" => {
    "App::Cmd" => 0,
    "Catmandu" => "1.0603",
    "Catmandu::DBI" => 0,
    "Catmandu::Importer::XML" => 0,
    "Catmandu::LIDO" => 0,
    "Catmandu::OAI" => 0,
    "Catmandu::Solr" => 0,
    "Catmandu::Store::Datahub" => 0,
    "Config::Onion" => 0,
    "Config::Simple" => 0,
    "DBI" => 0,
    "DateTime" => 0,
    "HTTP::Headers" => 0,
    "HTTP::Request::Common" => 0,
    "HTTP::Request::StreamingUpload" => 0,
    "JSON" => 0,
    "LWP::UserAgent" => 0,
    "Log::Any" => 0,
    "Log::Any::Adapter" => 0,
    "Log::Any::Adapter::Log4perl" => 0,
    "Log::Log4perl" => 0,
    "Module::Load" => 0,
    "Moo" => 0,
    "MooX::Aliases" => 0,
    "MooX::Role::Logger" => 0,
    "Moose::Role" => 0,
    "Ref::Util" => 0,
    "Sub::Exporter" => 0,
    "Term::ANSIColor" => 0,
    "Throwable::Error" => 0,
    "Try::Tiny" => 0,
    "Try::Tiny::ByClass" => 0,
    "URI::URL" => 0,
    "XML::LibXML" => 0,
    "namespace::clean" => 0,
    "perl" => "5.008005"
  },
  "script_files" => [
    "bin/dhconveyor"
  ],
  "test_requires" => {
    "Log::Any::Adapter::Log4perl" => 0,
    "Log::Any::Test" => "1.03",
    "Log::Log4perl" => 0,
    "Module::Build" => "0.28",
    "Test::Deep" => "0.112",
    "Test::Exception" => "0.43",
    "Test::LWP::UserAgent" => 0,
    "Test::More" => "0.96",
    "Test::Pod" => 0
  }
);


my %fallback_build_requires = (
  "Log::Any::Adapter::Log4perl" => 0,
  "Log::Any::Test" => "1.03",
  "Log::Log4perl" => 0,
  "Module::Build" => "0.28",
  "Test::Deep" => "0.112",
  "Test::Exception" => "0.43",
  "Test::LWP::UserAgent" => 0,
  "Test::More" => "0.96",
  "Test::Pod" => 0
);


unless ( eval { Module::Build->VERSION(0.4004) } ) {
  delete $module_build_args{test_requires};
  $module_build_args{build_requires} = \%fallback_build_requires;
}

my $build = Module::Build->new(%module_build_args);


$build->create_build_script;
